### PR TITLE
Add HNS API support

### PR DESCRIPTION
**Pull request by kurumiimari**

### CODE REVIEW

File: src/background/services/node.ts [hunk 1]
```diff
@@ -243,4 +213,24 @@ export default class NodeService extends GenericService {
   async stop() {
 
   }
+
+  async fetch(path: string|null, init: RequestInit): Promise<any> {
+    const { apiHost } = await this.exec('setting', 'getAPI');
+    const resp = await fetch(path ? `${apiHost}/${path}` : apiHost, init);
+
+    if (resp.status !== 200) {
```
Comment (Falci) [inline src/background/services/node.ts:221]:
```suggestion
    if (resp.status >= 400) {
```

File: src/background/services/node.ts [hunk 2]
```diff
@@ -243,4 +213,24 @@ export default class NodeService extends GenericService {
...
+      console.error(`Bad response code ${resp.status}.`);
+      
+      try {
+        const json = resp.json();
+        console.error('Body JSON:', json);
+      } catch (e) {
+        console.error('Error printing body JSON.');
```
Comment (Falci) [inline src/background/services/node.ts:228]:
```suggestion
        console.error('Error printing body JSON.', e);
```

File: src/background/services/wallet/index.ts [hunk 3]
```diff
@@ -1211,17 +1213,10 @@ export default class WalletService extends GenericService {
...
-      return transactions;
-    }
-
-    transactions = transactions.concat(newTXs);
-    return await this.getAllReceiveTXs(startDepth + ACCOUNT_DEPTH, endDepth + ACCOUNT_DEPTH, transactions);
+    return this.exec('node', 'getTXByAddresses', addresses, startDepth, endDepth);
   };
 
-  getAllChangeTXs = async (startDepth = 0, endDepth = ACCOUNT_DEPTH, transactions: any[] = []): Promise<any[]> => {
+  getAllChangeTXs = async (startDepth = 0, endDepth = 24): Promise<any[]> => {
```
Comment (Falci) [inline src/background/services/wallet/index.ts:1219]: `24` seems a magical number that appears multiple times. Please consider replacing by a constant with meaningful name.

File: src/background/services/node.ts [hunk 4]
```diff
@@ -34,18 +32,15 @@ export default class NodeService extends GenericService {
   }
 
   estimateSmartFee = async (opt: number) => {
-    const { apiHost } = await this.exec('setting', 'getAPI');
     const headers = await this.getHeaders();
-    const resp = await fetch(apiHost, {
+    return this.fetch(null,{
       method: 'POST',
       headers: headers,
```
Comment (Falci) [inline src/background/services/node.ts:38]:
```suggestion
```

File: src/background/services/node.ts [hunk 5]
```diff
@@ -66,49 +61,40 @@ export default class NodeService extends GenericService {
   };
 
   async getBlockchainInfo () {
-    const { apiHost } = await this.exec('setting', 'getAPI');
     const headers = await this.getHeaders();
 
-    const resp = await fetch(apiHost, {
+    return this.fetch(null, {
       method: 'POST',
       headers: headers,
```
Comment (Falci) [inline src/background/services/node.ts:68]:
```suggestion
```

File: src/background/services/node.ts [hunk 6]
```diff
@@ -66,49 +61,40 @@ export default class NodeService extends GenericService {
...
   }
 
   async sendRawTransaction (txJSON: any) {
-    const { apiHost } = await this.exec('setting', 'getAPI');
     const headers = await this.getHeaders();
 
-    const resp = await fetch(apiHost, {
+    return this.fetch(null, {
       method: 'POST',
       headers: headers,
```
Comment (Falci) [inline src/background/services/node.ts:81]:
```suggestion
```

File: src/background/services/node.ts [hunk 7]
```diff
@@ -127,9 +113,8 @@ export default class NodeService extends GenericService {
     const cachedEntry = await get(this.store, `namehash-${hash}`);
     if (cachedEntry) return cachedEntry;
 
-    const { apiHost } = await this.exec('setting', 'getAPI');
     const headers = await this.getHeaders();
-    const resp = await fetch(apiHost, {
+    const name = await this.fetch(null, {
       method: 'POST',
       headers: headers,
```
Comment (Falci) [inline src/background/services/node.ts:119]:
```suggestion
```

File: src/background/services/node.ts [hunk 8]
```diff
@@ -152,87 +136,73 @@ export default class NodeService extends GenericService {
   async getNameInfo(tld: string) {
     const { apiHost } = await this.exec('setting', 'getAPI');
     const headers = await this.getHeaders();
-    const resp = await fetch(apiHost, {
+    return fetch(apiHost, {
       method: 'POST',
       headers: headers,
```
Comment (Falci) [inline src/background/services/node.ts:141]:
```suggestion
```

File: src/background/services/node.ts [hunk 9]
```diff
@@ -152,87 +136,73 @@ export default class NodeService extends GenericService {
...
-    return json;
   }
 
   async getNameResource(tld: string) {
     const { apiHost } = await this.exec('setting', 'getAPI');
     const headers = await this.getHeaders();
-    const resp = await fetch(apiHost, {
+    return fetch(apiHost, {
       method: 'POST',
       headers: headers,
```
Comment (Falci) [inline src/background/services/node.ts:155]:
```suggestion
```

File: src/background/services/node.ts [hunk 10]
```diff
@@ -152,87 +136,73 @@ export default class NodeService extends GenericService {
...
   }
 
   async getTXByHash(txHash: string) {
-    const { apiHost } = await this.exec('setting', 'getAPI');
     const headers = await this.getHeaders();
-    const resp = await fetch(`${apiHost}/tx/${txHash}`, {
+    return this.fetch(`tx/${txHash}`, {
       method: 'GET',
       headers: headers,
     });
```
Comment (Falci) [inline src/background/services/node.ts:176]:
```suggestion
    return this.fetch(`tx/${txHash}`);
```

File: src/background/services/node.ts [hunk 11]
```diff
@@ -152,87 +136,73 @@ export default class NodeService extends GenericService {
...
   }
 
   async getCoin(txHash: string, txIndex: number) {
-    const { apiHost } = await this.exec('setting', 'getAPI');
     const headers = await this.getHeaders();
-    const resp = await fetch(`${apiHost}/coin/${txHash}/${txIndex}`, {
+    return this.fetch(`coin/${txHash}/${txIndex}`, {
       method: 'GET',
       headers: headers,
     });
```
Comment (Falci) [inline src/background/services/node.ts:168]:
```suggestion
    return this.fetch(`coin/${txHash}/${txIndex}`);
```

File: src/background/services/node.ts [hunk 12]
```diff
@@ -66,49 +61,40 @@ export default class NodeService extends GenericService {
...
     const cachedEntry = await get(this.store, `blockdata-${blockHeight}`);
     if (cachedEntry) return cachedEntry;
 
-    const { apiHost } = await this.exec('setting', 'getAPI');
     const headers = await this.getHeaders();
-    const resp = await fetch(`${apiHost}/block/${blockHeight}`, {
+    const block = await this.fetch(`block/${blockHeight}`, {
       method: 'GET',
       headers: headers,
     });
```
Comment (Falci) [inline src/background/services/node.ts:97]:
```suggestion
    const block = await this.fetch(`block/${blockHeight}`);
```

File: src/background/services/node.ts [hunk 13]
```diff
@@ -243,4 +213,24 @@ export default class NodeService extends GenericService {
   async stop() {
 
   }
+
+  async fetch(path: string|null, init: RequestInit): Promise<any> {
```
Comment (Falci) [inline src/background/services/node.ts:217]:
```suggestion
  async fetch(path: string|null, init: RequestInit = { method: 'GET', headers }): Promise<any> {
```

Comment (Falci) [reply src/background/services/node.ts:217]: Pretty sure this comment needs adjustments. The idea is to create a default header for all requests here.

**Comment by kurumiimari** [comment]
I'll take care of the comments this weekend. Thanks @Falci!

File: src/background/services/wallet/index.ts [hunk 14]
```diff
@@ -1386,47 +1380,13 @@ export default class WalletService extends GenericService {
...
         await pushMessage(setInfo(hash, height, time));
         this.emit('newBlock', {hash, height, time});
-      }, 1000);
-
-    });
-
-    socket.connect(apiHost);
-
-    this.socket = socket;
+    })(), 5000);
```
Comment (chikeichan) [inline src/background/services/wallet/index.ts:1389]: 5s poller might be too often, considering handshake blocks are like 15 minutes per block.

how about make this a minute? i think it's already manually checking when users open/close popup

File: src/background/services/wallet/index.ts [hunk 15]
```diff
@@ -1211,17 +1213,10 @@ export default class WalletService extends GenericService {
...
 
-    const newTXs = await this.exec('node', 'getTXByAddresses', addresses);
-
-    if (!newTXs.length) {
-      return transactions;
-    }
-
-    transactions = transactions.concat(newTXs);
-    return await this.getAllReceiveTXs(startDepth + ACCOUNT_DEPTH, endDepth + ACCOUNT_DEPTH, transactions);
+    return this.exec('node', 'getTXByAddresses', addresses, startDepth, endDepth);
```
Comment (chikeichan) [inline src/background/services/wallet/index.ts:1216]: i think the use of `startDepth` and `endDepth` here might be incorrect. The `startDepth` and `endDepth` variables indicates depth of the account, while the params passed to `getTXByAddresses` is intended for starting and ending block height to request tx from. Currently, this is requesting txs for addresses `[0 - 24]` for blocks `[0 - 24]`, and then addresses `[25 - 49]` for blocks `[25 - 49]`. 

I think we probably want to request blocks like this:
1. start with `addresses[0:99]` (to match bob desktop lookahead)
2. request all txs from `addresses[0:99]` for `blocks[0:n]` (`n = latest`)
  2a. if there are >100 txs (just my best guess, could be more), hapi returns 413
    - make `n = latest / 2`; repeat step 2;
  - if there are <= 100 txs, hapi returns 200
    - if `n < latest`; repeat step 2 with `blocks[latest-n:latest]`
3. repeat step 1 until all response from step 2 is empty

**Pull request merged**